File: medium-solution/Antidote21/142.java
public class Solution {
    public ListNode detectCycle(ListNode head) {
        Set<ListNode> set = new HashSet<ListNode>();
      
        while(head!=null){
            if(!set.add(head)){
                return head;
            }
            head = head.next;
        }
      
        return null;
        
    }
}

File: medium-solution/minesp3164/python/142.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def detectCycle(self, head):
        if not head:
            return None

        fast = head
        slow = head
        isCycle = False

        while(fast and fast.next):
            fast = fast.next.next
            slow = slow.next

            if(fast == slow):
                isCycle = True
                break

        if not isCycle:
            return None
        slow = head
        while fast != slow :
            slow = slow.next
            fast = fast.next
        return fast